### PR TITLE
Add Graal Feature for most common Caffeine usage

### PR DESCRIPTION
**Pull request by timyates**
This adds a Feature to allow reflective access to the most common Caffeine caches.

It is not a complete list, so documentation has been updated if people require further configuration.

Also, added a module which tests the native image, so we can catch errors early

### CODE REVIEW

**Comment by timyates** [comment]
Trying to find out why I'm getting:

```
> Task :test-suite-caffeine-native:nativeTestCompile SKIPPED
> Task :test-suite-caffeine-native:nativeTest SKIPPED
```

🤔

**Review by melix — CHANGES_REQUESTED** [review]
:x: I don't think this change works. I think `nativeTest` is skipped because the execution of `test`, in the _same_ build, led to no tests being executed or selected. Therefore, it's normal that `nativeTest` does nothing.

However, even if this wasn't the case, this would be broken, because the native test image builder wouldn't get any test ids to execute. So it would run, but execute no test.

**Comment by timyates** [comment]
@melix You mean this change?  https://github.com/micronaut-projects/micronaut-cache/pull/516/commits/2b64b4fec3603566883177aa83ed72f9643c92be

That just disables PTS so all (one of) the tests will be run every time, so nativeTest will be run every time?

Doesn't it?

**Comment by melix** [comment]
> @melix You mean this change? https://github.com/micronaut-projects/micronaut-cache/commit/2b64b4fec3603566883177aa83ed72f9643c92be

Yes

> That just disables PTS so all (one of) the tests will be run every time, so nativeTest will be run every time?
> Doesn't it?

Ah yes I misread the code, it disables PTS for test, so `nativeTest` will run. That said, I don't think we should do this. Either PTS works and we're good to go, or it doesn't and we have to disable it.

**Comment by timyates** [comment]
You're right...

PTS skipped the tests as the tests for cache-caffeine had previously passed, so no tests were required...

Apologies for the noise, it's more clever than I am ;-)

Removed the PTS disablement

File: src/main/docs/guide/cache-abstraction.adoc [hunk 1]
```diff
@@ -74,3 +74,10 @@ Names of caches under `micronaut.caches` should be defined in kebab case (lowerc
 ====
 
 To configure a weigher to be used with the `maximumWeight` configuration, create a bean that implements `com.github.benmanes.caffeine.cache.Weigher`. To associate a given weigher with only a specific cache, annotate the bean with `@Named(<cache name>)`. Weighers without a named qualifier will apply to all caches that don't have a named weigher. If no beans are found, a default implementation will be used.
+
+[NOTE]
+.Native compilation
+====
+When using Caffeine with Native Compilation, the most commonly used caches will be automatically registered.
+If you require additional caches, you will need to register them with Graal yourself https://docs.micronaut.io/latest/guide/#_adding_additional_classes_for_reflective_access[as shown in the guide].
```
Comment (alvarosanchez) [inline src/main/docs/guide/cache-abstraction.adoc:82]: How are users supposed to know if their cache is registered? Perhaps add a link to the feature class source?

**Review by alvarosanchez — APPROVED** [review]

**Review by melix — COMMENTED** [review]
Is there a way we can avoid the feature class and use the GraalVM reachability metadata repository, possibly with conditional metadata?

**Comment by timyates** [comment]
> Is there a way we can avoid the feature class and use the GraalVM reachability metadata repository, possibly with conditional metadata?

Do you have a link showing how all that works? Do our plugins support it? What would we need to do?

File: src/main/docs/guide/cache-abstraction.adoc (hunk 1 above)
Comment (timyates) [reply src/main/docs/guide/cache-abstraction.adoc:82]: The only way I can think of, is to look for ClassNotFound errors in Graal 🤔

Comment (alvarosanchez) [reply src/main/docs/guide/cache-abstraction.adoc:82]: Right, so then it's maybe a matter of adding a link to `CaffeineFeature.java`

**Comment by alvarosanchez** [comment]
> > Is there a way we can avoid the feature class and use the GraalVM reachability metadata repository, possibly with conditional metadata?
> 
> Do you have a link showing how all that works? Do our plugins support it? What would we need to do?

It is supported already in Native Build Tools, just not yet activated by default in micronaut-build.

As discussed the other day, I'm fine with this change today as a short/mid-term fix.

**Comment by timyates** [comment]
@melix Are we good now I've re-enabled PTS?

**Review by sdelamo — APPROVED** [review]

**Review by melix — APPROVED** [review]
So a quick question: do we need reflective access on all fields/methods? If not, it would be better to simply add the fields/methods which are strictly required.

**Comment by timyates** [comment]
@melix There are hundreds of classes which are generated dynamically in the Caffeine build...  I don't know how to find which are required 🤔

**Comment by timyates** [comment]
Ok, this is now waiting on https://github.com/micronaut-projects/micronaut-core/pull/8485

**Pull request merged**